### PR TITLE
Updating DataListItemAdd action to use doExecute().

### PR DESCRIPTION
**Pull request by karolkasas**
Updated the execute() function to doExecute() for the DataListItemAdd class.
Issue #2472815

### CODE REVIEW

File: src/Plugin/RulesAction/DataListItemAdd.php [hunk 1]
```diff
@@ -44,13 +44,21 @@
...
+   *   already in the list array.
+   * @param string $pos
+   *   Position in array where the item is being added.
    */
-  public function execute() {
-    $list = $this->getContextValue('list');
-    $item = $this->getContextValue('item');
-    $position = ($this->getContextValue('pos') ? $this->getContextValue('pos') : 'end');
-    $unique = ($this->getContextValue('unique') ? $this->getContextValue('unique') : FALSE);
+  public function doExecute($list, $item, $unique, $pos) {
```
Comment (klausi) [inline src/Plugin/RulesAction/DataListItemAdd.php]: this should be protected

File: src/Plugin/RulesAction/DataListItemAdd.php [hunk 2]
```diff
@@ -44,13 +44,21 @@
...
+    $position = ($pos ? $pos : 'end');
+    $unique = ($unique ? $unique : FALSE);
```
Comment (klausi) [inline src/Plugin/RulesAction/DataListItemAdd.php]: I think we can just use $unique = (bool) $unique;

File: src/Plugin/RulesAction/DataListItemAdd.php [hunk 3]
```diff
@@ -44,13 +44,21 @@
...
     // Optionally, only add the list item if it is not yet contained.
     if (!((bool) $unique && in_array($item, $list))) {
```
Comment (klausi) [inline src/Plugin/RulesAction/DataListItemAdd.php:64]: Then we don't need the (bool) cast here

**Comment by karolkasas** [comment]
The test are passing without that line of code just fine, I have also fixed the protected access for the function. @klausi

File: src/Plugin/RulesAction/DataListItemAdd.php [hunk 4]
```diff
@@ -44,13 +44,20 @@
...
+   *
+   * @param array $list
+   *   An array to add an item to.
+   * @param mixed $item
+   *   An item which is being added to the array.
+   * @param bool $unique
+   *   A boolean value which indicates if the item should be updated if it is
+   *   already in the list array.
+   * @param string $pos
+   *   Position in array where the item is being added.
```
Comment (fago) [inline src/Plugin/RulesAction/DataListItemAdd.php]: This should document the available options as well that it it's optional. Also is unique not optional also?

File: src/Plugin/RulesAction/DataListItemAdd.php [hunk 5]
```diff
@@ -44,13 +44,20 @@
...
    */
-  public function execute() {
-    $list = $this->getContextValue('list');
-    $item = $this->getContextValue('item');
-    $position = ($this->getContextValue('pos') ? $this->getContextValue('pos') : 'end');
-    $unique = ($this->getContextValue('unique') ? $this->getContextValue('unique') : FALSE);
+  protected function doExecute($list, $item, $unique, $pos) {
```
Comment (fago) [inline src/Plugin/RulesAction/DataListItemAdd.php]: should repeat the default values here

Comment (karolkasas) [inline src/Plugin/RulesAction/DataListItemAdd.php]: Eh you are right, I forgot about the default values. Will fix this when I will be behind com.

File: src/Plugin/RulesAction/DataListItemAdd.php (hunk 4 above)
Comment (karolkasas) [inline src/Plugin/RulesAction/DataListItemAdd.php]: Sorry I can't follow, you mean that the param comments should contain that the last two params are optional?

Comment (ndeet) [inline src/Plugin/RulesAction/DataListItemAdd.php]: I think he means that in the description of the param you should prefix the optional params with e.g. `(optional) Position in array ...`

**Comment by karolkasas** [comment]
I have finally found time and fixed the optional params and comments for the params.

**Comment by karolkasas** [comment]
@klausi @fago does this need additional work still?

File: src/Plugin/RulesAction/DataListItemAdd.php [hunk 6]
```diff
@@ -44,16 +44,25 @@
...
+   *   An array to add an item to.
+   * @param mixed $item
+   *   An item which is being added to the array.
+   * @param bool $unique
+   *   (optional) A boolean value which indicates if the item should be updated
+   *   if it is already in the list array.
+   * @param string $pos
+   *   (optional) Position in array where the item is being added. Accepted
+   *   value are:
+   *   - "start"      Beginning of the array.
```
Comment (klausi) [inline src/Plugin/RulesAction/DataListItemAdd.php:59]: can we use a colon after the value example similar to https://www.drupal.org/coding-standards/docs#lists and without the additional spacing?

File: src/Plugin/RulesAction/DataListItemAdd.php [hunk 7]
```diff
@@ -44,16 +44,25 @@
...
+   *   - "end"        End of the array.
```
Comment (klausi) [inline src/Plugin/RulesAction/DataListItemAdd.php:60]: same here

**Comment by klausi** [comment]
continued in #326

**Pull request closed without merging**